sdks/python/apache_beam/runners/portability/universal_local_runner.py

```diff
@@ -0,0 +1,398 @@
+
+from apache_beam.portability.api import beam_fn_api_pb2
+from apache_beam.portability.api import beam_job_api_pb2
+from apache_beam.portability.api import beam_job_api_pb2_grpc
+from apache_beam.runners import runner
+from apache_beam.runners.portability import fn_api_runner
+
+TERMINAL_STATES = [
+    beam_job_api_pb2.JobState.DONE,
+    beam_job_api_pb2.JobState.STOPPED,
+    beam_job_api_pb2.JobState.FAILED,
+    beam_job_api_pb2.JobState.CANCELLED,
+]
+
+
+class UniversalLocalRunner(runner.PipelineRunner):
+  """A BeamRunner that executes pipelines via the Beam Job API.
+
+  By default, this runner executes in process but still uses GRPC to communicate
+  pipeline and worker state.  It can also be configured to use inline calls
+  rather than GRPC (for speed) or launch completely separate subprocesses for
+  the runner and worker(s).
+  """
+
+  def __init__(self, use_grpc=True, use_subprocesses=False):
+    super(UniversalLocalRunner, self).__init__()
+    self._use_grpc = use_grpc
+    self._use_subprocesses = use_subprocesses
+
+    self._handle = None
+    self._subprocess = None
+
+  def __del__(self):
+    # Best effort to not leave any dangling processes around.
+    self.cleanup()
+
+  def cleanup(self):
+    if self._subprocess:
+      self._subprocess.kill()
+      time.sleep(0.1)
+    self._subprocess = None
+
+  def _get_handle(self):
+    if not self._handle:
+      if self._use_subprocesses:
+        if self._subprocess:
+          # Kill the old one if it exists.
+          self._subprocess.kill()
+        # TODO(robertwb): Consider letting the subprocess pick one and
+        # communicate it back...
+        port = _pick_unused_port()
+        logging.info("Starting server on port %d.", port)
+        self._subprocess = subprocess.Popen([
+            sys.executable,
+            '-m',
+            'apache_beam.runners.portability.universal_local_runner_main',
+            '-p',
+            str(port),
+            '--worker_command_line',
+            '%s -m apache_beam.runners.worker.sdk_worker_main' % sys.executable
+        ])
+        handle = beam_job_api_pb2_grpc.JobServiceStub(
+            grpc.insecure_channel('localhost:%d' % port))
+        logging.info("Waiting for server to be ready...")
+        start = time.time()
+        timeout = 30
+        while True:
+          time.sleep(0.1)
+          if self._subprocess.poll() is not None:
+            raise RuntimeError(
+                "Subprocess terminated unexpectedly with exit code %d." %
+                self._subprocess.returncode)
+          elif time.time() - start > timeout:
+            raise RuntimeError(
+                "Pipeline timed out waiting for job service subprocess.")
+          else:
+            try:
+              handle.GetState(
+                  beam_job_api_pb2.GetJobStateRequest(job_id='[fake]'))
+              break
+            except grpc.RpcError as exn:
+              if exn.code != grpc.StatusCode.UNAVAILABLE:
```

Comment:
<!--new_thread; commit:77e51af3ee2af39264c366a2602957e7e9bb20c7; resolved:0-->
Why do we ignore other RPC errors, as opposed to wait until we  receive grpc.StatusCode.OK?

Reply:
We can't issue a valid request at this point (well, I didn't want to create a bunch of fake jobs that would build up) so we get a "real" error once our get state fails. Added a comment.